{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "460455d4",
   "metadata": {},
   "outputs": [],
   "source": [
    "dna_codon_table = {\n",
    "    \"TTT\": \"F\", \"TTC\": \"F\", \"TTA\": \"L\", \"TTG\": \"L\",\n",
    "    \"TCT\": \"S\", \"TCC\": \"S\", \"TCA\": \"S\", \"TCG\": \"S\",\n",
    "    \"TAT\": \"Y\", \"TAC\": \"Y\", \"TAA\": \"*\", \"TAG\": \"*\",\n",
    "    \"TGT\": \"C\", \"TGC\": \"C\", \"TGA\": \"*\", \"TGG\": \"W\",\n",
    "    \"CTT\": \"L\", \"CTC\": \"L\", \"CTA\": \"L\", \"CTG\": \"L\",\n",
    "    \"CCT\": \"P\", \"CCC\": \"P\", \"CCA\": \"P\", \"CCG\": \"P\",\n",
    "    \"CAT\": \"H\", \"CAC\": \"H\", \"CAA\": \"Q\", \"CAG\": \"Q\",\n",
    "    \"CGT\": \"R\", \"CGC\": \"R\", \"CGA\": \"R\", \"CGG\": \"R\",\n",
    "    \"ATT\": \"I\", \"ATC\": \"I\", \"ATA\": \"I\", \"ATG\": \"M\",\n",
    "    \"ACT\": \"T\", \"ACC\": \"T\", \"ACA\": \"T\", \"ACG\": \"T\",\n",
    "    \"AAT\": \"N\", \"AAC\": \"N\", \"AAA\": \"K\", \"AAG\": \"K\",\n",
    "    \"AGT\": \"S\", \"AGC\": \"S\", \"AGA\": \"R\", \"AGG\": \"R\",\n",
    "    \"GTT\": \"V\", \"GTC\": \"V\", \"GTA\": \"V\", \"GTG\": \"V\",\n",
    "    \"GCT\": \"A\", \"GCC\": \"A\", \"GCA\": \"A\", \"GCG\": \"A\",\n",
    "    \"GAT\": \"D\", \"GAC\": \"D\", \"GAA\": \"E\", \"GAG\": \"E\",\n",
    "    \"GGT\": \"G\", \"GGC\": \"G\", \"GGA\": \"G\", \"GGG\": \"G\",\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "55c89482",
   "metadata": {},
   "outputs": [],
   "source": [
    "rna_codon_table = {\n",
    "    \"UUU\": \"F\", \"UUC\": \"F\", \"UUA\": \"L\", \"UUG\": \"L\",\n",
    "    \"UCU\": \"S\", \"UCC\": \"S\", \"UCA\": \"S\", \"UCG\": \"S\",\n",
    "    \"UAU\": \"Y\", \"UAC\": \"Y\", \"UAA\": \"*\", \"UAG\": \"*\",\n",
    "    \"UGU\": \"C\", \"UGC\": \"C\", \"UGA\": \"*\", \"UGG\": \"W\",\n",
    "    \"CUU\": \"L\", \"CUC\": \"L\", \"CUA\": \"L\", \"CUG\": \"L\",\n",
    "    \"CCU\": \"P\", \"CCC\": \"P\", \"CCA\": \"P\", \"CCG\": \"P\",\n",
    "    \"CAU\": \"H\", \"CAC\": \"H\", \"CAA\": \"Q\", \"CAG\": \"Q\",\n",
    "    \"CGU\": \"R\", \"CGC\": \"R\", \"CGA\": \"R\", \"CGG\": \"R\",\n",
    "    \"AUU\": \"I\", \"AUC\": \"I\", \"AUA\": \"I\", \"AUG\": \"M\",\n",
    "    \"ACU\": \"T\", \"ACC\": \"T\", \"ACA\": \"T\", \"ACG\": \"T\",\n",
    "    \"AAU\": \"N\", \"AAC\": \"N\", \"AAA\": \"K\", \"AAG\": \"K\",\n",
    "    \"AGU\": \"S\", \"AGC\": \"S\", \"AGA\": \"R\", \"AGG\": \"R\",\n",
    "    \"GUU\": \"V\", \"GUC\": \"V\", \"GUA\": \"V\", \"GUG\": \"V\",\n",
    "    \"GCU\": \"A\", \"GCC\": \"A\", \"GCA\": \"A\", \"GCG\": \"A\",\n",
    "    \"GAU\": \"D\", \"GAC\": \"D\", \"GAA\": \"E\", \"GAG\": \"E\",\n",
    "    \"GGU\": \"G\", \"GGC\": \"G\", \"GGA\": \"G\", \"GGG\": \"G\",\n",
    "}"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "03887276",
   "metadata": {},
   "source": [
    "## Simple Translator"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "d3914536",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Enter the type of sequence (1 for DNA, 2 for RNA): 1\n",
      "Enter a DNA sequence: atgatcaaacga\n",
      "Amino Acid Sequence: MIKR\n",
      "Do you want to translate again? (y/n): n\n"
     ]
    }
   ],
   "source": [
    "def valid_sequence(sequence, seq_type):\n",
    "    valid_nucleotides = \"ATGC\" if seq_type == \"DNA\" else \"AUGC\"\n",
    "    return all(nucleotide in valid_nucleotides for nucleotide in sequence)\n",
    "\n",
    "def translate(sequence, seq_type):\n",
    "    codon_table = dna_codon_table if seq_type == \"DNA\" else rna_codon_table\n",
    "    \n",
    "    if not valid_sequence(sequence, seq_type):\n",
    "        print(\"Invalid input: The sequence should contain only 'A', 'T', 'G', 'C' for DNA or 'A', 'U', 'G', 'C' for RNA.\")\n",
    "        return None\n",
    "    \n",
    "    if seq_type == \"RNA\":\n",
    "        sequence = sequence.replace(\"U\", \"T\")\n",
    "    \n",
    "    aa_sequence = \"\"\n",
    "    \n",
    "    for i in range(0, len(sequence), 3):\n",
    "        codon = sequence[i:i + 3]\n",
    "        if codon in codon_table:\n",
    "            aa_sequence += codon_table[codon]\n",
    "        else:\n",
    "            aa_sequence += \"- Codon does not exist -\"\n",
    "    \n",
    "    return aa_sequence\n",
    "\n",
    "while True:\n",
    "    seq_type = input(\"Enter the type of sequence (1 for DNA, 2 for RNA): \")\n",
    "    if seq_type == \"1\":\n",
    "        codon_table = dna_codon_table\n",
    "        seq_type = \"DNA\"\n",
    "    elif seq_type == \"2\":\n",
    "        codon_table = rna_codon_table\n",
    "        seq_type = \"RNA\"\n",
    "    else:\n",
    "        print(\"Invalid input: Please enter 1 for DNA or 2 for RNA.\")\n",
    "        continue\n",
    "    \n",
    "    sequence = input(f\"Enter a {seq_type} sequence: \").upper()\n",
    "    \n",
    "    aa_sequence = translate(sequence, seq_type)\n",
    "    if aa_sequence:\n",
    "        print(\"Amino Acid Sequence:\", aa_sequence)\n",
    "    \n",
    "    choice = input(\"Do you want to translate again? (y/n): \").lower()\n",
    "    if choice != \"y\":\n",
    "        break\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "cba9f472",
   "metadata": {},
   "source": [
    "## Translates all Frames"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "f1a9b66f",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Enter the type of sequence (1 for DNA, 2 for RNA): 1\n",
      "Enter a DNA sequence: atgttcatcgttag\n",
      "Frame 1 (5' to 3'): MFIV-\n",
      "Frame 2 (5' to 3'): CSSL-\n",
      "Frame 3 (5' to 3'): VHR*\n",
      "Frame 1 (3' to 5'): LTMN-\n",
      "Frame 2 (3' to 5'): *R*T-\n",
      "Frame 3 (3' to 5'): NDEH\n",
      "Do you want to translate again? (y/n): y\n",
      "Enter the type of sequence (1 for DNA, 2 for RNA): 2\n",
      "Enter a RNA sequence: augcagaaaa\n",
      "Frame 1 (5' to 3'): -QK-\n",
      "Frame 2 (5' to 3'): -RK\n",
      "Frame 3 (5' to 3'): AE-\n",
      "Frame 1 (3' to 5'): FSA-\n",
      "Frame 2 (3' to 5'): FLH\n",
      "Frame 3 (3' to 5'): FC-\n",
      "Do you want to translate again? (y/n): n\n"
     ]
    }
   ],
   "source": [
    "def valid_sequence(sequence, seq_type):\n",
    "  valid_nucleotides = \"ATGC\" if seq_type == \"1\" else \"AUGC\"\n",
    "  return all(nucleotide in valid_nucleotides for nucleotide in sequence)\n",
    "\n",
    "def translate(sequence, seq_type, codon_table):\n",
    "  sequence = sequence.upper() \n",
    "     \n",
    "  translated_frames = []\n",
    "   \n",
    "  for frame in range(3):\n",
    "    translated_frames.append([])\n",
    "    for i in range(frame, len(sequence), 3):\n",
    "      codon = sequence[i:i + 3]\n",
    "      if len(codon) == 3 and codon in codon_table: \n",
    "        translated_frames[frame].append(codon_table[codon])\n",
    "      else:\n",
    "        translated_frames[frame].append(\"-\")\n",
    "\n",
    "  reverse_complement = sequence[::-1].replace(\"U\", \"T\").translate(str.maketrans(\"ATGC\", \"TACG\"))\n",
    "\n",
    "  for frame in range(3):\n",
    "    translated_frames.append([])\n",
    "    for i in range(frame, len(reverse_complement), 3):\n",
    "      codon = reverse_complement[i:i + 3]\n",
    "      if len(codon) == 3 and codon in codon_table: \n",
    "        translated_frames[frame + 3].append(codon_table[codon])\n",
    "      else:\n",
    "        translated_frames[frame + 3].append(\"-\")\n",
    "\n",
    "  return translated_frames\n",
    "\n",
    "\n",
    "while True:\n",
    "  seq_type = input(\"Enter the type of sequence (1 for DNA, 2 for RNA): \")\n",
    "  if seq_type == \"1\":\n",
    "    codon_table = dna_codon_table\n",
    "    seq_type = \"DNA\"\n",
    "  elif seq_type == \"2\":\n",
    "    codon_table = dna_codon_table\n",
    "    seq_type = \"RNA\"\n",
    "  else:\n",
    "    print(\"Invalid input: Please enter 1 for DNA or 2 for RNA.\")\n",
    "    continue\n",
    "   \n",
    "  sequence = input(f\"Enter a {seq_type} sequence: \").strip()\n",
    "   \n",
    "  translation_result = translate(sequence, seq_type, codon_table)\n",
    "   \n",
    "  for i, frame in enumerate(translation_result, 1):\n",
    "    direction = \"5' to 3'\" if i <= 3 else \"3' to 5'\"\n",
    "    frame_number = i if i <= 3 else i - 3\n",
    "    print(f\"Frame {frame_number} ({direction}): {''.join(frame)}\")\n",
    "   \n",
    "  choice = input(\"Do you want to translate again? (y/n): \").strip().lower()\n",
    "  if choice != \"y\":\n",
    "    break"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "6f171b90",
   "metadata": {},
   "source": [
    "## Translates with Start and Stop Codon"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "id": "4c77a1d5",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Enter the type of sequence (1 for DNA, 2 for RNA): 1\n",
      "Enter a DNA sequence: CTCAGTCAGGCGCTCAGCTCCGTTTCGGTTTCACTTCCGGTGGAGGGCCGCCTCTGAGCGGGCGGCGGGC CGACGGCGAGCGCGGGCGGCGGCGGTGACGGAGGCGCCGCTGCCAGGGGGCGTGCGGCAGCGCGGCGGCG GCGGCGGCGGCGGCGGCGGCGGAGGCGGCGGCGGCGGCGGCGGCGGCGGCGGCTGGGCCTCGAGCGCCCG CAGCCCACCTCTCGGGGGCGGGCTCCCGGCGCTAGCAGGGCTGAAGAGAAGATGGAGGAGCTGGTGGTGG AAGTGCGGGGCTCCAATGGCGCTTTCTACAAGGTACTTGGCTCTAGGGCAGGCCCCATCTTCGCCCTTCC TTCCCTCCCTTTTCTTCTTGGTGTCGGCGGGAGGCAGGCCCGGGGCCCTCTTCCCGAGCACCGCGCCTGG GTGCCAGGGCACGCTCGGCGGGATGTTGTTGGGAGGGAAGGACTGGACTTGGGGCCTGTTGGAAGCCCCT CTCCGACTCCGAGAGGCCCTAGCGCCTATCGAAATGAGAGACCAGCGAGGAGAGGGTTCTCTTTCGGCGC CGAGCCCCGCCGGGGTGAGCTGGGGATGGGCGAGGGCCGGCGGCAGGTACTAGAGCCGGGCGGGAAGGGC CGAAATCGGCGCTAAGTGACGGCGATGGCTTATTCCCCCTTTCCTAAACATCATCTCCCAGCGGGATCCG GGCCTGTCGTGTGGGTAGTTGTGGAGGAGCGGGGGGCGCTTCAGCCGGGCCGCCTCCTGCAGCGCCAAGA GGGCTTCAGGTCTCCTTTGGCTTCTCTTTTCCGGTCTAGCATTGGGACTTCGGAGAGCTCCACTGTTCTG GGCGAGGGCTGTGAAGAAAGAGTAGTAAGAAGCGGTAGTCGGCACCAAATCACAATGGCAACTGATTTTT AGTGGCTTCTCTTTGTGGATTTCGGAGGAGATTTTAGATCCAAAAGTTTCAGGAAGACCCTAACATGGCC\n",
      "Frame 1 (5' to 3'): MEELVV-SAGLQWRFLQGTWL\n",
      "Frame 2 (5' to 3'): MLLGGKDWTWGLLEAP-PTPRGPSAYRNERPARRGFSFG-RAPPG\n",
      "Frame 3 (5' to 3'): MALSTRYLALGQAPSSPF-SLPFLLGVGGRQARGPLPEHRAW-ARARSAGCCWEGRTGLGACWKP-LRLREALAPIEMRDQRGEGSLSA-EPRRGELGMGEGRRQVLEPGGKG-KSALSDGDGLFPLS\n",
      "Frame 1 (3' to 5'): MMFRKGGISHRRHLAPIS-PSRPALVPAAGPRPSPAHPGGA-APKENPLLAGLSFR\n",
      "Frame 2 (3' to 5'): MLGSS\n",
      "Frame 3 (3' to 5'): MLDRKREAKGDLKP-LALQEAARLKRPPLLHNYPHDR-RIPLGDDV\n",
      "Do you want to translate again? (y/n): n\n"
     ]
    }
   ],
   "source": [
    "def valid_sequence(sequence, seq_type):\n",
    "    valid_nucleotides = \"ATGC\" if seq_type == \"1\" else \"AUGC\"\n",
    "    return all(nucleotide in valid_nucleotides for nucleotide in sequence)\n",
    "\n",
    "def find_start_codon(frame, sequence):\n",
    "    start_codons = [\"ATG\", \"AUG\"]\n",
    "    for i in range(frame, len(sequence), 3):\n",
    "        codon = sequence[i:i + 3]\n",
    "        if codon in start_codons:\n",
    "            return i\n",
    "    return None\n",
    "\n",
    "def translate(sequence, seq_type, codon_table):\n",
    "    sequence = sequence.upper()\n",
    "\n",
    "    translated_frames = []\n",
    "\n",
    "    for frame in range(3):\n",
    "        start_index = find_start_codon(frame, sequence)\n",
    "        if start_index is not None:\n",
    "            translated_frames.append([])\n",
    "            for i in range(start_index, len(sequence), 3):\n",
    "                codon = sequence[i:i + 3]\n",
    "                if len(codon) == 3 and codon in codon_table:\n",
    "                    if codon_table[codon] == \"*\":\n",
    "                        break  \n",
    "                    translated_frames[frame].append(codon_table[codon])\n",
    "                else:\n",
    "                    translated_frames[frame].append(\"-\")\n",
    "        else:\n",
    "            translated_frames.append([\"No start codon found\" if all(codon not in sequence for codon in [\"ATG\", \"AUG\"]) else \"Start codon not in frame\"])\n",
    "\n",
    "    reverse_complement = sequence[::-1].replace(\"U\", \"T\").translate(str.maketrans(\"ATGC\", \"TACG\"))\n",
    "\n",
    "    for frame in range(3):\n",
    "        start_index = find_start_codon(frame, reverse_complement)\n",
    "        if start_index is not None:\n",
    "            translated_frames.append([])\n",
    "            for i in range(start_index, len(reverse_complement), 3):\n",
    "                codon = reverse_complement[i:i + 3]\n",
    "                if len(codon) == 3 and codon in codon_table:\n",
    "                    if codon_table[codon] == \"*\":\n",
    "                        break \n",
    "                    translated_frames[frame + 3].append(codon_table[codon])\n",
    "                else:\n",
    "                    translated_frames[frame + 3].append(\"-\")\n",
    "        else:\n",
    "            translated_frames.append([\"No start codon found\" if all(codon not in reverse_complement for codon in [\"ATG\", \"AUG\"]) else \"Start codon not in frame\"])\n",
    "\n",
    "    return translated_frames\n",
    "\n",
    "while True:\n",
    "    seq_type = input(\"Enter the type of sequence (1 for DNA, 2 for RNA): \")\n",
    "    if seq_type == \"1\":\n",
    "        codon_table = dna_codon_table\n",
    "        seq_type = \"DNA\"\n",
    "    elif seq_type == \"2\":\n",
    "        codon_table = rna_codon_table\n",
    "        seq_type = \"RNA\"\n",
    "    else:\n",
    "        print(\"Invalid input: Please enter 1 for DNA or 2 for RNA.\")\n",
    "        continue\n",
    "\n",
    "    sequence = input(f\"Enter a {seq_type} sequence: \").strip()\n",
    "\n",
    "    translation_result = translate(sequence, seq_type, codon_table)\n",
    "\n",
    "    for i, frame in enumerate(translation_result, 1):\n",
    "        direction = \"5' to 3'\" if i <= 3 else \"3' to 5'\"\n",
    "        frame_number = i if i <= 3 else i - 3\n",
    "        print(f\"Frame {frame_number} ({direction}): {''.join(frame)}\")\n",
    "\n",
    "    choice = input(\"Do you want to translate again? (y/n): \").strip().lower()\n",
    "    if choice != \"y\":\n",
    "        break\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "00a1486d",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.6"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
